navigation/navigation-compose-material/src/main/java/androidx/navigation/compose/material/BottomSheetNavigator.kt

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.navigation.compose.material
+
+import androidx.compose.foundation.layout.ColumnScope
+import androidx.compose.material.ExperimentalMaterialApi
+import androidx.compose.material.ModalBottomSheetState
+import androidx.compose.material.ModalBottomSheetValue
+import androidx.compose.material.rememberModalBottomSheetState
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.collectAsState
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.lifecycle.Lifecycle
+import androidx.navigation.FloatingWindow
+import androidx.navigation.NavBackStackEntry
+import androidx.navigation.NavDestination
+import androidx.navigation.Navigator
+import androidx.navigation.NavigatorState
+import androidx.navigation.compose.material.BottomSheetNavigator.Destination
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+
+/**
+ * Create and remember a [BottomSheetNavigator]
+ *
+ * @param sheetState The [ModalBottomSheetState] that the [BottomSheetNavigator] will use to
+ * drive the sheet state
+ */
+@OptIn(ExperimentalMaterialApi::class)
```

Comment:
We'll need to follow up the Compose team - while we can use experimental APIs in alpha, policy states that we **cannot** move to beta with APIs that depend on experimental APIs from another artifact group. We'll need to get agreement that these APIs will become stable before Compose 1.0 before we can add APIs built on top of them.

I can do most of this work if you'd like - I'd appreciate a list of APIs we'll need to implement this fully (or if there's some minimal set we can use and leave helpers that need more experimental APIs as demos / docs only).

Reply:
Oh, interesting. That makes sense and is probably for the better 😄 

I think there are quite a few issues around bottom sheets and the swipeable APIs and I'm not sure what the plan is, but my understanding was that these might well stay experimental until after 1.0. Looking at the APIs we're using though - I think it might be possible to make them stable. At least `ModalBottomSheetState` can probably be made stable by moving the `SwipeableState` into the class and not inheriting from it anymore (same as with `DrawerState`, https://android-review.googlesource.com/c/platform/frameworks/support/+/1613082/).

If you could talk to the team about this, that'd be awesome! (You can also make a group chat with Matvei, you and me if you want)

The APIs we need are
- `ModalBottomSheetLayout` (though we only need that for our convenience method that's `@ExperimentalMaterialApi` anyways)
- `ModalBottomSheetState` for use in `BottomSheetNavigator` and `SheetContentHost`

The APIs we need from `ModalBottomSheetState` are all related to visibility but that's pretty much at the core of the swipeable issues anyways.